### PR TITLE
[3.5] feat: add wal write system call metrics observation

### PR DESCRIPTION
**Pull request by Akiqqqqqqq**
xref. #17615 

add new metrics: _**wal_write_duration_seconds**_

**_wal_write_duration_seconds_** represent total delay of: 
> syscall.write(padding_info + record_data_with_padding)

- delay of fsync is not included in _**wal_write_duration_seconds**_
- data: the raft log to write to the WAL on disk at one single raftNode loop

### CODE REVIEW

File: tools/mod/go.sum [hunk 1]
```diff
@@ -138,6 +138,7 @@ github.com/golang/protobuf v1.4.0-rc.4.0.20200313231945-b860323f09d0/go.mod h1:W
 github.com/golang/protobuf v1.4.0/go.mod h1:jodUvKwWbYaEsadDk5Fwe5c77LiNKVO9IDvqG2KuDX0=
 github.com/golang/protobuf v1.4.1 h1:ZFgWrT+bLgsYPirOnRfKLYJLvssAegOj/hgyMFdJZe0=
 github.com/golang/protobuf v1.4.1/go.mod h1:U8fpvMrcmy5pZrNK1lt4xCsGvpyWQ/VVv6QDs8UjoX8=
+github.com/golang/protobuf v1.5.4/go.mod h1:lnTiLA8Wa4RWRcIUkrtSVa5nRhsEGBg48fD6rSs7xps=
```
Comment (jmhbnz) [inline tools/mod/go.sum:141]: Hey @Akiqqqqqqq - Thanks for proposing this. This change looks unrelated, is it intentional?

Comment (Akiqqqqqqq) [reply tools/mod/go.sum:141]: sry,  it is not intentional , will remove this later

**Comment by liangyuanpeng** [comment]
Seem like  the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of `ok-to-test` but not approve this PR to run github action.

Remind me if i missed something.

https://files.slack.com/files-pri/T6XGWV83B-F06QKHKHVAM/image.png

**Comment by jmhbnz** [comment]
> Seem like the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of `ok-to-test` but not approve this PR to run github action.

Very strange. I have not seen this issue before. Can you please rebase and force push to trigger the workflows again?

**Comment by Akiqqqqqqq** [comment]
> > Seem like the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of `ok-to-test` but not approve this PR to run github action.
> 
> Very strange. I have not seen this issue before. Can you please rebase and force push to trigger the workflows again?

to be sure, so do I need to perfrom again?
`git rebase HEAD~1 --signoff`
and
`git push -f`

**Comment by Akiqqqqqqq** [comment]
> > Seem like the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of `ok-to-test` but not approve this PR to run github action.
> 
> Very strange. I have not seen this issue before. Can you please rebase and force push to trigger the workflows again?

I retried, but seems to be not working

**Comment by jmhbnz** [comment]
> I retried, but seems to be not working

Closing and re-opening the pr seems to have fixed it 😅

**Review by jmhbnz — APPROVED** [review]
LGTM - Reading through the linked issue I think this addition makes sense, thanks @Akiqqqqqqq for drafting the pr.

Edit: Note if we do go ahead with merging this we need to update docs etc for the new metric.

File: server/wal/encoder.go [hunk 2]
```diff
@@ -93,7 +94,9 @@ func (e *encoder) encode(rec *walpb.Record) error {
 		data = append(data, make([]byte, padBytes)...)
 	}
 	n, err = e.bw.Write(data)
+	start := time.Now()
```
Comment (serathius) [inline server/wal/encoder.go:97]: Should this line be before the Write?

Comment (Akiqqqqqqq) [reply server/wal/encoder.go:96]: oh, my mistake, will fix soon

**Review by liangyuanpeng — COMMENTED** [review]
> Seem like the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of ok-to-test but not approve this PR to run github action.

This caught my attention, so I looked into it a little deeper. I found out that this PR was being submitted for `release-3.5` instead of `main`, and the `release-3.5` branch didn't have the workflow of  https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml , so this is the reason the github acton workflow wasn't being run by approve .

Should we submit a PR to the main branch first?

**Comment by Akiqqqqqqq** [comment]
> > Seem like the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of ok-to-test but not approve this PR to run github action.
> 
> This caught my attention, so I looked into it a little deeper. I found out that this PR was being submitted for `release-3.5` instead of `main`, and the `release-3.5` branch didn't have the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml , so this is the reason the github acton workflow wasn't being run by approve .
> 
> I think we should create a PR to the `main` first?

**Comment by Akiqqqqqqq** [comment]
> > Seem like the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of ok-to-test but not approve this PR to run github action.
> 
> This caught my attention, so I looked into it a little deeper. I found out that this PR was being submitted for `release-3.5` instead of `main`, and the `release-3.5` branch didn't have the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml , so this is the reason the github acton workflow wasn't being run by approve .
> 
> I think we should create a PR to the `main` first?

ok， let me raise another PR to main

**Comment by Akiqqqqqqq** [comment]
> > Seem like the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml have some problem here? Exist label of ok-to-test but not approve this PR to run github action.
> 
> This caught my attention, so I looked into it a little deeper. I found out that this PR was being submitted for `release-3.5` instead of `main`, and the `release-3.5` branch didn't have the workflow of https://github.com/etcd-io/etcd/blob/main/.github/workflows/gh-workflow-approve.yaml , so this is the reason the github acton workflow wasn't being run by approve .
> 
> I think we should create a PR to the `main` first?

check this out: https://github.com/etcd-io/etcd/pull/17618

**Comment by jmhbnz** [comment]
> I think we should create a PR to the `main` first?

Yes good catch - let's merge #17618 first.

**Comment by ivanvc** [comment]
/retitle [3.5] feat: add wal write system call metrics observation

**Review by fuweid — APPROVED** [review]

**Review by ahrtr — APPROVED** [review]

**Comment by Akiqqqqqqq** [comment]
> #17615

got it

**Review by serathius — APPROVED** [review]
